# INDENG 242 HW1 Ankit Bhawsar 3038588568 Q4


## Import Necessary Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.simplefilter("ignore", UserWarning)
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.stats.outliers_influence import variance_inflation_factor

## Import Dataset and Print Head

In [2]:
rogue_data = pd.read_csv("Rogue_242.csv")
rogue_data.info()
rogue_data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 162 entries, 0 to 161
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   MonthNumeric  162 non-null    int64  
 1   MonthFactor   162 non-null    object 
 2   Year          162 non-null    int64  
 3   RogueSales    162 non-null    int64  
 4   Unemployment  162 non-null    float64
 5   RogueQueries  162 non-null    int64  
 6   CPIAll        162 non-null    float64
 7   CPIEnergy     162 non-null    float64
dtypes: float64(3), int64(4), object(1)
memory usage: 10.2+ KB


MonthNumeric MonthFactor  Year  RogueSales  Unemployment  RogueQueries  \
0             1     January  2008        5435           5.0            18   
1             2    February  2008        5223           4.9            11   
2             3       March  2008        6873           5.1            18   
3             4       April  2008        5814           5.0            17   
4             5         May  2008        7467           5.4            17   

    CPIAll  CPIEnergy  
0  212.174    226.775  
1  212.687    229.731  
2  213.448    233.349  
3  213.942    234.778  
4  215.208    243.924

## Convert Categorical Variables into Binary Dummy Variables

Store column names in col variable.


In [3]:
rogue_data = pd.get_dummies(rogue_data, columns = ['MonthFactor'],drop_first = True)
rogue_data.drop(columns='MonthNumeric',inplace=True)

In [4]:
rogue_data.head()

Year  RogueSales  Unemployment  RogueQueries   CPIAll  CPIEnergy  \
0  2008        5435           5.0            18  212.174    226.775   
1  2008        5223           4.9            11  212.687    229.731   
2  2008        6873           5.1            18  213.448    233.349   
3  2008        5814           5.0            17  213.942    234.778   
4  2008        7467           5.4            17  215.208    243.924   

   MonthFactor_August  MonthFactor_Decemeber  MonthFactor_February  \
0                   0                      0                     0   
1                   0                      0                     1   
2                   0                      0                     0   
3                   0                      0                     0   
4                   0                      0                     0   

   MonthFactor_January  MonthFactor_July  MonthFactor_June  MonthFactor_March  \
0                    1                 0                 0                  0   
1                    0                 0                 0                  0   
2                    0                 0                 0                  1   
3                    0                 0                 0                  0   
4                    0                 0                 0                  0   

   MonthFactor_May  MonthFactor_November  MonthFactor_October  \
0                0                     0                    0   
1                0                     0                    0   
2                0                     0                    0   
3                0                     0                    0   
4                1                     0                    0   

   MonthFactor_Septeber  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0

## Split Dataset into Train and Test in ratio 75%-25% while preserving chronological order
For this problem we have to split the dataset into training and testing set in chronological order, and then populate X_train, y_train, X_test, y_test variables.

## Q 4 (a) (i)

**It is important to maintain chronological order when using time-series data because it is possible that the more recent data is more important to the model than less recent data. i.e. the latest data gets more weightage than the older data.**

**Applying this to our case, we want our model to predict Nissan Rogue sales for the future which means more recent data, say from the 2020s is more relevant to us than say data from 2008. This is why we dont split the data randomly**

In [5]:
rogue_train = rogue_data[rogue_data['Year']<=2017]
rogue_test = rogue_data[rogue_data['Year']>2017]

print("Train Shape= ",rogue_train.shape)
print("Test Shape= ",rogue_test.shape)

rogue_test.head()

Train Shape=  (120, 17)
Test Shape=  (42, 17)


Year  RogueSales  Unemployment  RogueQueries   CPIAll  CPIEnergy  \
120  2018       36184           4.0            72  248.721    217.273   
121  2018       38119           4.1            72  249.300    219.765   
122  2018       42151           4.0            71  249.517    216.492   
123  2018       23331           4.0            72  250.275    219.424   
124  2018       38413           3.8            79  250.786    220.770   

     MonthFactor_August  MonthFactor_Decemeber  MonthFactor_February  \
120                   0                      0                     0   
121                   0                      0                     1   
122                   0                      0                     0   
123                   0                      0                     0   
124                   0                      0                     0   

     MonthFactor_January  MonthFactor_July  MonthFactor_June  \
120                    1                 0                 0   
121                    0                 0                 0   
122                    0                 0                 0   
123                    0                 0                 0   
124                    0                 0                 0   

     MonthFactor_March  MonthFactor_May  MonthFactor_November  \
120                  0                0                     0   
121                  0                0                     0   
122                  1                0                     0   
123                  0                0                     0   
124                  0                1                     0   

     MonthFactor_October  MonthFactor_Septeber  
120                    0                     0  
121                    0                     0  
122                    0                     0  
123                    0                     0  
124                    0                     0

In [6]:
X_train = rogue_train.drop(columns=["RogueSales"])
X_test = rogue_test.drop(columns=["RogueSales"])

y_train = rogue_train['RogueSales']
y_test = rogue_test['RogueSales']

print("X Train Shape= ",X_train.shape)
print("y Train Shape= ",y_train.shape)
print("X Test Shape= ",X_test.shape)
print("y Test Shape= ",y_test.shape)

y_test.reset_index(drop=True,inplace=True)
X_test.reset_index(drop=True,inplace=True)


X Train Shape=  (120, 16)
y Train Shape=  (120,)
X Test Shape=  (42, 16)
y Test Shape=  (42,)


### Fit Simple Linear Regression Model to Data


In [7]:
simple_model = LinearRegression()
simple_model.fit(X=X_train,y=y_train)

R_squared = simple_model.score(X_train, y_train)
print('R_squared:', R_squared)
print()

intercept = simple_model.intercept_
coefficients = simple_model.coef_
features = simple_model.feature_names_in_

print('Features:', features)
print()
print('Intercept:', intercept)
print()
print('Coefficients:', coefficients)

R_squared: 0.9033906803790014

Features: ['Year' 'Unemployment' 'RogueQueries' 'CPIAll' 'CPIEnergy'
 'MonthFactor_August' 'MonthFactor_Decemeber' 'MonthFactor_February'
 'MonthFactor_January' 'MonthFactor_July' 'MonthFactor_June'
 'MonthFactor_March' 'MonthFactor_May' 'MonthFactor_November'
 'MonthFactor_October' 'MonthFactor_Septeber']

Intercept: -109149.49947215345

Coefficients: [   20.17119267  -174.52546053   230.11703206   401.65281232
   -71.35843825  2951.29263128  3210.7673643   2450.4934264
 -1112.07651339  1799.43947741  1321.2679001   4656.84536934
  2801.40527045   722.36306006    90.79395546  1706.47435034]


## Q 4 (a) (ii)

**Above we see the intercept and different coefficients assigned by the model while fitting the data.**

**We know the linear regression model is as follows:**

**Z = intercept + coeff_1(x1) + coeff_2(x2) + ... + coeff_p(xp)**

**We can rewrite the above equation for our model as:**

**Z = -109149.49947215345 + 20.17119267(x1) -174.52546053(x2) + 230.11703206(x3) + ... + 1706.47435034(x16)**

**or**

**Z = -109149.49947215345 + 20.17119267(Year) -174.52546053(Unemployment) + 230.11703206(CPIEnergy) + ... + 1706.47435034(MonthFactor_Septeber)**

**This means that the dependent variable Z will change by a factor coefficient(i) if independent variable (i) changes by 1.**

**ex: if values of Unemployment increases or decreases by 1, RogueSales will decrease or increase by a factor of 174 respectively.**

---




### Finding Predictions of X_test data

In [8]:
y_pred = simple_model.predict(X_test)
y_pred

array([30709.44247814, 34309.29162211, 36623.69390769, 32292.19546122,
       36848.52117744, 35453.72422995, 36875.74662076, 37538.83516005,
       33685.12104266, 32489.89588217, 34168.50785297, 37140.05110784,
       33235.18270509, 37361.39130827, 41136.07341381, 35290.73535165,
       39713.983669  , 38921.10449708, 40221.96783883, 40014.52290478,
       38526.00198035, 35259.10874754, 37789.01889725, 39216.5023818 ,
       36089.03675364, 40534.64858357, 37968.63215901, 29154.01200294,
       37477.50252152, 40262.01378231, 42065.06162822, 42386.49776435,
       39473.82622957, 39898.9538307 , 41801.52412159, 45097.86041171,
       40397.31685372, 44921.74566304, 48188.4239337 , 44349.36582655,
       46975.09982756, 47133.87322867])


### Function to calculate Out-of-Sample R^2 Value

We write our own function to find Out of Sample R^2 value.

In [9]:
def OutOfSample_r2(y_test,y_pred,y_train):
    
    avg = np.mean(y_train)
    RSS = np.sum((y_test - y_pred)**2)
    TSS = np.sum((y_test - avg )**2)

    return 1-(RSS/TSS)


In [10]:
print(OutOfSample_r2(y_test,y_pred,y_train))

0.06652804670167667


## Q 4 (a) (iii)

**<ins>Out of Sample R2</ins>**

**OSR^2 is a metric similar to R^2 but based on unseen (test) data. This gives us an estimate of how close our model fits to previously unseen data. If R^2 value is high but OSR^2 value is low it means our model has overfitted the training the data and producers a very high error on testing data.**

**Formula for OSR^2 is given as:**

**OSR^2 = 1 - (RSS/TSS)**

**where RSS is the residual sum of squares, which is the sum of squares of the deviations of predicted test values (y_pred) from actual test values (y_test)**

**where TSS is total sum of squares, which is the sum of squares of deviations of actual test values (y_test) from the output of our baseline model, which in our case predicts the average of training data.**

**We can observe that our Out-of-Sample R^2 value is only 0.06 which means our model is performing very badly. This is because our model is trained on data split keeping the chronological constant. The model is not robust enough to predict the future sales of cars. In the next sections, we use random sampling to produce much better performance.

---



## Now Splitting the Dataset Randomly

In [11]:
rogue_train, rogue_test = train_test_split(rogue_data, test_size = 0.25, random_state = 4040)

X_train = rogue_train.drop(columns=["RogueSales"])
X_test = rogue_test.drop(columns=["RogueSales"])
y_train = rogue_train['RogueSales']
y_test = rogue_test['RogueSales']


X_train.reset_index(drop=True,inplace=True)
y_train.reset_index(drop=True,inplace=True)
X_test.reset_index(drop=True,inplace=True)
y_test.reset_index(drop=True,inplace=True)


print(X_train['Year'].head())
print(X_test['Year'].head())

0    2016
1    2019
2    2016
3    2010
4    2013
Name: Year, dtype: int64
0    2014
1    2012
2    2017
3    2009
4    2012
Name: Year, dtype: int64


In [12]:
simple_model = LinearRegression()
simple_model.fit(X=X_train,y=y_train)

R_squared = simple_model.score(X_train, y_train)
print('R_squared:', R_squared)
print()

intercept = simple_model.intercept_
coefficients = simple_model.coef_
features = simple_model.feature_names_in_

print('Features:', features)
print()
print('Intercept:', intercept)
print()
print('Coefficients:', coefficients)

R_squared: 0.8508501554635496

Features: ['Year' 'Unemployment' 'RogueQueries' 'CPIAll' 'CPIEnergy'
 'MonthFactor_August' 'MonthFactor_Decemeber' 'MonthFactor_February'
 'MonthFactor_January' 'MonthFactor_July' 'MonthFactor_June'
 'MonthFactor_March' 'MonthFactor_May' 'MonthFactor_November'
 'MonthFactor_October' 'MonthFactor_Septeber']

Intercept: -13939453.362845039

Coefficients: [ 7116.73147492 -1348.27898317   205.13106456 -1707.92323989
   109.25530042  5884.3178172   7830.30712773  2299.75444868
 -3825.90080336  2587.25788059  4075.04960516  3151.89092469
  3401.27166118  3566.85358299  3445.19554033  4011.57920585]


In [13]:
y_pred = simple_model.predict(X_test)
y_pred

array([23201.75701819,  8792.89625369, 31620.73742402,  5005.76283776,
       11379.5111842 , 12062.97131286, 10172.55231177, 22656.49070006,
       26603.02418649, 31877.22819295, 27502.26247391, 28009.82042625,
        9995.37170922, 23761.78285549, 26375.17598934, 27631.58455396,
        7436.67694237, 31930.84956139, 28670.91125873, 19084.51234805,
        7713.30966697, 29686.42515455, 32390.97419478, 27906.67108793,
       27950.45594922, 32498.78736091, 18240.63841612, 28158.36756012,
       29355.92577103, 33566.82277402, 12519.81574141, 15161.51177488,
       11853.28144753, 13029.35037271, 10958.17212092,  7136.68432578,
       21479.29477861, 12631.60838542, 27421.74315641, 31660.54685949,
       10739.32722487])

In [14]:
print(OutOfSample_r2(y_test,y_pred,y_train))

0.8416318359772264



---

### Calculate VIF for each explanatory variable

## Q4 (b) (i)


**Variance Inflation Factor:**

**VIF is a metric used to identify colliniarity between the independent features of a model. Linear regression models assume independent variables are completely independent from each other. This is so that the actual effect of each feature on the dependent variable can be captured.**

**VIF of a variable (i) is defined as the the ratio of overall variance and variance of a model that contains only variable (i). So a very high VIF means that variable is highly collinear.**

**Formula: VIF = 1/(1 - r(i)^2)**

**Where r(i)^2 is the r^2 value of regressing the ith variable on other variables.**

**So if r^2 == 0, there is no relation between variable and other variables,  VIF = 1, i.e. there is no collinearity.**

**Practically any VIF from 1-5 is acceptable.**


In [15]:
def VIF(df, columns):
    
    values = sm.add_constant(df[columns]).values
    num_columns = len(columns)+1
    vif = [variance_inflation_factor(values, i) for i in range(num_columns)]
    
    return pd.Series(vif[1:], index=columns)


### Function to calculate Out-of-Sample R^2 Value
VIF(X_train,['Year','Unemployment','RogueQueries','CPIAll','CPIEnergy'])

Year            81.242317
Unemployment     1.706625
RogueQueries    16.647399
CPIAll          91.997053
CPIEnergy        2.716457
dtype: float64



## Q4 (b) (ii)


**Since VIF of some variables is very high, we try dropping them**


In [16]:
print(VIF(X_train,['Year','Unemployment','RogueQueries','CPIEnergy']))

Year            11.929895
Unemployment     1.646086
RogueQueries    13.837601
CPIEnergy        1.102315
dtype: float64


In [17]:
print(VIF(X_train,['Year','Unemployment','CPIEnergy']))

Year            1.402770
Unemployment    1.296848
CPIEnergy       1.095436
dtype: float64


In [18]:
X_train = rogue_train.drop(columns=["RogueSales"])
X_test = rogue_test.drop(columns=["RogueSales"])

y_train = rogue_train['RogueSales']
y_test = rogue_test['RogueSales']

X_train = X_train.drop(columns=['RogueQueries','CPIAll'])
X_test = X_test.drop(columns=['RogueQueries','CPIAll'])


In [19]:
simple_model = LinearRegression()
simple_model.fit(X=X_train,y=y_train)

R_squared = simple_model.score(X_train, y_train)
print('R_squared:', R_squared)
print()

intercept = simple_model.intercept_
coefficients = simple_model.coef_
features = simple_model.feature_names_in_

print('Features:', features)
print()
print('Intercept:', intercept)
print()
print('Coefficients:', coefficients)

R_squared: 0.8089941136468248

Features: ['Year' 'Unemployment' 'CPIEnergy' 'MonthFactor_August'
 'MonthFactor_Decemeber' 'MonthFactor_February' 'MonthFactor_January'
 'MonthFactor_July' 'MonthFactor_June' 'MonthFactor_March'
 'MonthFactor_May' 'MonthFactor_November' 'MonthFactor_October'
 'MonthFactor_Septeber']

Intercept: -2853777.2079107673

Coefficients: [ 1435.73487035 -1807.04548265   -44.47771681  4811.81029675
  3454.9147893   3879.16322683 -1690.09754024  2936.46309123
  3430.43194339  4230.75691726  3146.71261028   267.38554311
   124.7771686   1032.26880732]


In [20]:
y_pred = simple_model.predict(X_test)
y_pred

array([18448.68609375,  9564.12201311, 28667.00736703,  4918.31171838,
       12651.47000732, 12495.3120953 ,  8658.48236764, 23143.79081477,
       24953.67402048, 30231.39176899, 25547.31648543, 26548.09999186,
        6582.36994144, 21435.83790831, 26876.39541216, 26709.1096694 ,
        7197.94896486, 30909.2655748 , 27990.47258048, 16775.15337474,
        9312.47214126, 26752.20902199, 29996.35469184, 27093.32100161,
       25545.55919359, 32717.57610954, 18462.03688303, 27843.01735386,
       26116.58531978, 31617.28600552, 13216.55358494, 15255.4899886 ,
       13814.40596814, 13041.85025703,  9431.12841137,  6842.77120952,
       21408.29874005,  8597.0494746 , 27754.46797389, 34445.1485122 ,
       13985.14175365])

In [21]:
print(OutOfSample_r2(y_test,y_pred,y_train))

0.7776575428457726


## Q4 (b) (iii)

**The two variables were dropped because they had extremely high collinearity. For example, the RogueQueries column was dropped because it is obvious that it will have collinearity with the Year column because internet usage in 2008 will be significantly lower than that in 2022. This means these two columns provide basically the same information to the model and one of them can be dropped.**

**As we can see the OSR^2 value of this new improved model (0.77) is slightly lower than that using all features (0.84). But now we have a model that is perfectly suited to use all the information provided without worrying about one variable skewing the data of another variable.**

**We can also see from the display below that:**

**Coefficient of 'Year' has decreased, meaning its significance has decreased, coefficient of 'Unemployment' has decreased, meaning its significance has decreased,etc**
